cosim/black-parrot-example/ps.cpp

```diff
@@ -382,31 +389,42 @@ void nbf_load(bp_zynq_pl *zpl, char *nbf_filename) {
       i++;
     }
     nbf[i] = std::stoull(nbf_command, nullptr, 16);
-    if (nbf[0] == 0x3) {
-      // we map BP physical addresses for DRAM (0x8000_0000 - 0x9FFF_FFFF)
-      // (256MB)
+
+    if (nbf[0] == 0x3 || nbf[0] == 0x2 || nbf[0] == 0x1 || nbf[0] == 0x0) {
+      // we map BP physical addresses for DRAM (0x8000_0000 - 0x9FFF_FFFF) (256MB)
       // to the same ARM physical addresses
       // see top_fpga.v for more details
 
-      if (nbf[1] >= 0x80000000) {
-        address = nbf[1];
-        address = address;
-        data = nbf[2];
-        nbf[2] = nbf[2] >> 32;
-        zpl->axil_write(address, data, 0xf);
-        address = address + 4;
-        data = nbf[2];
-        zpl->axil_write(address, data, 0xf);
-      }
       // we map BP physical address for CSRs etc (0x0000_0000 - 0x0FFF_FFFF)
       // to ARM address to 0xA0000_0000 - 0xAFFF_FFFF  (256MB)
+      if (nbf[1] >= 0x80000000)
+        base_addr = GP1_ADDR_BASE - 0x80000000;
+      else
+        base_addr = GP1_ADDR_BASE + 0x20000000;
+
+      if (nbf[0] == 0x3) {
+        zpl->axil_write(base_addr + nbf[1], nbf[2], 0xf);
+        zpl->axil_write(base_addr + nbf[1] + 4, nbf[2] >> 32, 0xf);
+      }
+      else if (nbf[0] == 0x2) {
+        zpl->axil_write(base_addr + nbf[1], nbf[2], 0xf);
+      }
+      else if (nbf[0] == 0x1) {
+        int offset = nbf[1] % 4;
+        int shift = 2 * offset;
+        data = zpl->axil_read(base_addr + nbf[1] - offset);
+        data = data & rotl((uint32_t)0xffff0000,shift) + nbf[2] & ((uint32_t)0x0000ffff << shift);
+        zpl->axil_write(base_addr + nbf[1] - offset, data, 0xf);
+      }
       else {
-        address = nbf[1];
-        address = address + 0xA0000000;
-        data = nbf[2];
-        zpl->axil_write(address, data, 0xf);
+        int offset = nbf[1] % 4;
```

Comment:
Seems like we should implement the wmask instead of requiring a RMW? Fine with raising an issue and not blocking this